lib/connector/OAuthConnector.js

```diff
@@ -182,52 +227,90 @@ class OAuthConnector {
+     * foreignVendorId entry in the userContext.foreignOAuthIdentities rather than a user of this connector.
      */
-    async ensureAccessToken(fusebitContext, userContext) {
-        if (userContext.vendorToken.access_token && userContext.vendorToken.expires_at > Date.now() + 30000) {
-            Sdk.debug('RETURNING CURRENT ACCESS TOKEN FOR USER', userContext.vendorUserId);
-            return userContext.vendorToken;
-        }
-        if (userContext.vendorToken.refresh_token) {
-            Sdk.debug('REFRESHING ACCESS TOKEN FOR USER', userContext.vendorUserId);
-            userContext.status = 'refreshing';
-            try {
-                await this.saveUser(fusebitContext, userContext);
-                userContext.vendorToken = await this.refreshAccessToken(fusebitContext, userContext.vendorToken);
-                if (!isNaN(userContext.vendorToken.expires_in)) {
-                    userContext.vendorToken.expires_at = Date.now() + +userContext.vendorToken.expires_in * 1000;
-                }
-                userContext.vendorUserProfile = await this.getUserProfile(userContext.vendorToken);
-                userContext.status = 'authenticated';
-                await this.saveUser(fusebitContext, userContext);
+    async ensureAccessToken(fusebitContext, userContext, foreignVendorId) {
+        if (foreignVendorId) {
+            // Get access token from foreign OAuth connector specified in userContext.foreignOAuthIdentities
+            const oauthIdentity = (userContext.foreignOAuthIdentities || {})[foreignVendorId];
+            Sdk.debug('OBTAINING ACCESS TOKEN FOR FOREIGN USER', foreignVendorId, oauthIdentity);
+            if (oauthIdentity) {
+                const response = await Superagent.get(
+                    `${oauthIdentity.connectorBaseUrl}/user/${encodeURIComponent(oauthIdentity.userId)}/token`
+                ).set('Authorization', `Bearer ${fusebitContext.fusebit.functionAccessToken}`);
+                return response.body;
+            } else {
+                throw new Error(
+                    `The user ${userContext.vendorUserId} is not associated with an identity in the ${foreignVendorId} OAuth connector.`
+                );
+            }
+        } else {
+            // Get access token for "this" OAuth connector
+            if (userContext.vendorToken.access_token && userContext.vendorToken.expires_at > Date.now() + 30000) {
+                Sdk.debug('RETURNING CURRENT ACCESS TOKEN FOR USER', userContext.vendorUserId);
                 return userContext.vendorToken;
-            } catch (e) {
-                await this.deleteUser(fusebitContext, userContext.vendorUserId);
-                Sdk.debug('REFRESH TOKEN ERROR', e);
+            }
+            if (userContext.vendorToken.refresh_token) {
+                Sdk.debug('REFRESHING ACCESS TOKEN FOR USER', userContext.vendorUserId);
+                userContext.status = 'refreshing';
```

Comment:
Where are these used?  Is there a check that keeps a stampeding herd from occurring?

Reply:
See my previous comment on the exponential backoff - just added this logic.

Reply:
Is there some logic here already that does a 2-phase commit to prevent multiple lambda’s from simultaneously starting a refresh cycle?

Reply:
No. There is a little race so a few lambdas can start the refresh simultaneously until one of them saves the user context with status = 'refreshing'. The last one to complete will write out the user context with access token.

Reply:
Any coordination activity like this is going to be problematic, due to scaling behavior of lambda.  Additionally, degenerate conditions will cause the remote service to blacklist the app, bringing down all of the customers tied to it.

This is guaranteed to penalize when operating at scale.  Normal fixes would include serialization through a queue, a two phase commit through storage, or a singleton somewhere managing all OAuth requests.

If this code is still largely seen as transitory (that is, it’s all going to go away in v2) then this is probably fine for now, but if there’s an expectation that the code will live longer then some type of locking should be applied now.

Reply:
Of course. It is better than nothing though. Solving it right really means addressing the overall throttling problem space.